cluster/selfmanaged.go

```
@@ -0,0 +1,141 @@
+func (s *SelfManaged) Receive(c *actor.Context) {
+	switch msg := c.Message().(type) {
+	case actor.Started:
+		s.pid = c.PID()
+		s.members.Add(s.cluster.Member())
+		members := &Members{
+			Members: s.members.Slice(),
+		}
+		s.cluster.engine.Send(s.cluster.PID(), members)
+		s.memberPinger = c.SendRepeat(c.PID(), memberPing{}, memberPingInterval)
+		s.start(c)
+	case actor.Stopped:
+		s.memberPinger.Stop()
+		s.cluster.engine.Unsubscribe(s.eventSubPID)
+	case *MembersJoin:
+		for _, member := range msg.Members {
+			s.addMember(member)
+		}
+		ourMembers := &Members{
+			Members: s.members.Slice(),
+		}
+		s.members.ForEach(func(member *Member) bool {
+			s.cluster.engine.Send(memberToProviderPID(member), ourMembers)
+			return true
+		})
+	case *Members:
+		for _, member := range msg.Members {
+			if !s.members.Contains(member) {
+				s.addMember(member)
+			}
+		}
+		if s.members.Len() > 0 {
+			members := &Members{
+				Members: s.members.Slice(),
+			}
+			s.cluster.engine.Send(s.cluster.PID(), members)
+		}
+	case memberPing:
+		s.members.ForEach(func(member *Member) bool {
+			if member.Host != s.cluster.agentPID.Address {
+				ping := &actor.Ping{
+					From: c.PID(),
+				}
+				c.Send(memberToProviderPID(member), ping)
+			}
+			return true
+		})
+	case memberLeave:
+		member := s.members.GetByHost(msg.ListenAddr)
+		s.removeMember(member)
+	}
+}
+
+// If we receive members from another node in the cluster
+// we respond with all the members we know of, and ofcourse
+// add the new one.
+func (s *SelfManaged) addMember(member *Member) {
+	if !s.members.Contains(member) {
+		s.members.Add(member)
+	}
+}
+
+func (s *SelfManaged) removeMember(member *Member) {
+	if s.members.Contains(member) {
+		s.members.Remove(member)
+	}
+	s.updateCluster()
+}
+
+func (s *SelfManaged) updateCluster() {
+	members := &Members{
+		Members: s.members.Slice(),
+	}
+	s.cluster.engine.Send(s.cluster.PID(), members)
+}
+
+func (s *SelfManaged) start(c *actor.Context) error {
```

does this need to return an error?